### PR TITLE
Merge `diff_local` into `profile_local`

### PR DESCRIPTION
**Pull request by nnethercote**
The `diff_local` subcommand has a lot in common with `profile_local`,
but also has some unnecessary differences in behaviour, such as not
using an `<ID>` argument. This commit removes `diff_local` and moves the
diff functionality to `profile_local` by adding a single new option to
it, which makes things more consistent and simplifies the code.

The commit also changes the naming of output files, to be more
consistent. For example:

**OLD**

`collector diff_local cachegrind $RUSTC1 $RUSTC2 --builds Check --runs Full`

raw output
- `cgout-after-helloworld-Check-Full`
- `cgout-before-helloworld-Check-Full`

filtered raw output
- `cgfilt-after-helloworld-Check-Full`
- `cgfilt-before-helloworld-Check-Full`

human-readable output
- `cgann-after-helloworld-Check-Full`
- `cgann-before-helloworld-Check-Full`

filtered raw diff
- `cgdiff-before-after-helloworld-Check-Full`

human-readable diff
- `cgann-before-after-helloworld-Check-Full`
- `cgdiffann-latest`

**NEW**

`collector profile_local cachegrind $RUSTC1 MyId --rustc2 $RUSTC2 --builds Check --runs Full`

raw output
- `cgout-MyId1-helloworld-Check-Full`
- `cgout-MyId2-helloworld-Check-Full`

filtered raw output
- `cgfilt-MyId1-helloworld-Check-Full`
- `cgfilt-MyId2-helloworld-Check-Full`

human-readable output
- `cgann-MyId1-helloworld-Check-Full`
- `cgann-MyId2-helloworld-Check-Full`

filtered raw diff
- `cgfilt-diff-MyId1-MyId2-helloworld-Check-Full`

human-readable diff
- `cgann-diff-MyId1-MyId2-helloworld-Check-Full`
- `cgann-diff-latest`

The diff files are named more consistently and the prefixes are more
consistent.

Also, with `diff_local` if you specified a toolchain via a `+abcdef`
revid it would use `abcdef` as the Id instead of `before` or `after`.
When diffing with `profile_local` it now uses `MyId1`/`MyId2`, for
consistency with the non-diff case.

Some of the "is this binary installed?" checks are moved around
so they are only run if necessary.

Finally, this commit documents the new behaviour in the README. (There
was no documentation at all for `diff_local`.) It also documents how
`+`-prefixed `<RUSTC>` options work, something that was also missing
from the docs.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
This looks pretty good to me, but I'm not sure I'm a fan of the ID being MyId for the self profile page. There was a certain appeal to the use of commit hashes there.

I think it may make sense to default IDs to the toolchain name (when specified with +syntax) - basically bringing that from diff local to profile local. But I suppose that *could* be done separately, it just feels like this would be a regression otherwise.

I think this default would basically 100% match my workflow, both when using CI artifacts and local rustcs... but not sure if it does others.

**Comment by nnethercote** [comment]
I use non `+syntax` forms all the time. I only learned about `+syntax` recently :)

> I think it may make sense to default IDs to the toolchain name (when specified with +syntax) - basically bringing that from diff local to profile local. But I suppose that _could_ be done separately, it just feels like this would be a regression otherwise.

The downside of that is that if you use `+syntax` for all the compilers being tested (be it one or two) then the `ID` is mandatory but isn't used.

When you use `+syntax` what form does it usually take -- a 40-digit revid? Or something like `+nightly`?

**Comment by Mark-Simulacrum** [comment]
Right, my goal would be to make the ID be optional (or optional with the +syntax).

I typically have two ways of using the +syntax form:

1. diff_local with two 40-digit revids, typically pulled directly from the detailed-query/self-profile page (first step on investigating perf results)
  * auto-generated, so not *too* important what the exact syntax here is, in principle. But, nice if we avoid mangling artifacts from other commits -- sometimes useful, though pretty rarely.
2. diff_local and/or profile_local with +stage1 and +stage1-rust2 (roughly), pointing to different local directories where I build rustc (with rustup toolchain link).

Typically for (2) I'm not too worried about clobbering outputs, because I'm tracking the "key" statistic in a separate empty text editor (e.g., instruction counts for the whole run in cachegrind). I'll actually often run diff_local and wait 2x as long (for baseline and 'next' toolchains) today just to get the automated pipeline producing the cgdiffann-latest file, mostly because I'm "lazy" and reproducing the right flags and sed's and such can be a bit of a pain. OTOH, the in-tree filtering fails to remove directory differences from cachegrind results (it expects a commit hash but I have `/rust`, `/rust-2`, `/rust-3`). I'll sometimes edit in place and just not commit that, since it's easier than writing a shell script to do that.


I don't think I've ever really given truly meaningful IDs. The only pair I will sometimes use is "baseline" and then just "next" while iterating, mostly since they're short and easier to type if I'm trying to look at e.g. results/eprintln-baseline...

**Comment by nnethercote** [comment]
Interesting. I have a script for doing cachegrind diffs, it does the filtering of uninteresting differences. It would be good to get that into `profile_local`.

As for IDs, I often end up with a sequence (e.g. `Leb0`, `Leb1`, `Leb2`, `Leb3`) of them when I'm trying out a bunch of things. And I also use a text file to record key stats, but I find it useful to have the whole sequence of files available to look over.

I never use rustup toolchain link. I have env vars `$RUSTC0`, `$RUSTC1`, `$RUSTC2`, `RUSTC3` which point to the built binaries in the various repos I have.

**Comment by nnethercote** [comment]
Here's a proposal that should satisfy all our current use cases:
- `<ID>` becomes non-mandatory, i.e. `--id <ID>`
- You can specify the compiler via a path or a +toolchain
- If path: by default, use `Path` as the ID in the filename
- If +toolchain: by default, use the toolchain name as the ID in the filename
- If `--id` is specified, use that in the filename instead (appending `1` and `2` for diffs)

As a follow-up, we could consider removing the ability to specify the compiler via path. That would mandate the use of `rustup toolchain link` on local compiler builds.

**Comment by nnethercote** [comment]
I have updated the code to add `--id` as I described above. See the commit message for more details.

File: collector/README.md [hunk 1]
```diff
@@ -384,23 +392,29 @@ The mandatory `<PROFILER>` argument must be one of the following.
...
 
 ### Profiling options
 
 The following options alter the behaviour of the `profile_local` subcommand.
 - `--builds <BUILDS>`: as for `bench_local`.
 - `--cargo <CARGO>`: as for `bench_local`.
 - `--exclude <EXCLUDE>`: as for `bench_local`.
+- `--id <ID>`: an identifer that will form part of the output filenames.
+  Also, if `--rustc2` is identified, a `1` will be appended to the ID for the
+  first run and a `2` will be appended to the ID for the second run.
```
Comment (Mark-Simulacrum) [inline collector/README.md:409]: Do you think it's worth clarifying that this appending only happens with `--id` or path-based rustc's, not with the +-based synthesis?

As I think about this, it may make sense to actually change that to appending `-1` or `-2` if the ID would otherwise be the same before and after. This would let us easily e.g. do trivial comparisons for noise by passing identical toolchains before/after. It might be easier to explain, too.

**Review by Mark-Simulacrum — APPROVED** [review]
Happy to merge with the suggestion or we can leave it as-is, no strong opinion.

File: collector/README.md (hunk 1 above)
Comment (nnethercote) [reply collector/README.md:409]: It's worth clarifying, yes.

I don't want to append `-1` and `-2` if the toolchains are the same; that's an additional edge case and thus makes it harder to explain. I think I'll error out in that case. If anyone wants to test the same toolchain twice they can use `--id` and get different IDs that way.
